File: php-laravel/app/Models/Tool.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Tool extends Model
{
    protected $table = 'tools';
    
    protected $fillable = [
        'name',
        'description',
        'vendor',
        'website_url',
        'category_id',
        'monthly_cost',
        'active_users_count',
        'owner_department',
        'status'
    ];

    protected $casts = [
        'monthly_cost' => 'decimal:2',
        'active_users_count' => 'integer',
        'created_at' => 'datetime',
        'updated_at' => 'datetime'
    ];

    /**
     * Get the category that owns the tool
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class, 'category_id');
    }

    /**
     * Scope for active tools only
     */
    public function scopeActive($query)
    {
        return $query->where('status', 'active');
    }

    /**
     * Scope for filtering by department
     */
    public function scopeDepartment($query, $department)
    {
        return $query->where('owner_department', $department);
    }

    /**
     * Scope for filtering by category
     */
    public function scopeCategory($query, $categoryId)
    {
        return $query->where('category_id', $categoryId);
    }

    /**
     * Scope for filtering by cost range
     */
    public function scopeCostBetween($query, $min, $max)
    {
        return $query->whereBetween('monthly_cost', [$min, $max]);
    }
}

File: python-fastapi/app/schemas/tool.py
"""Pydantic schemas for request/response validation."""
from datetime import datetime
from decimal import Decimal
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field, HttpUrl, ConfigDict

from app.models import DepartmentType, ToolStatusType


# Tool Schemas
class ToolBase(BaseModel):
    """Base tool schema."""
    name: str = Field(..., min_length=2, max_length=100, description="Tool name")
    description: Optional[str] = Field(None, description="Tool description")
    vendor: str = Field(..., max_length=100, description="Vendor name")
    website_url: Optional[HttpUrl] = Field(None, description="Tool website URL")
    category_id: int = Field(..., gt=0, description="Category ID")
    monthly_cost: Decimal = Field(..., ge=0, max_digits=10, decimal_places=2, description="Monthly cost per user")
    owner_department: DepartmentType = Field(..., description="Owner department")


class ToolCreate(ToolBase):
    """Schema for creating a tool."""
    pass


class ToolUpdate(BaseModel):
    """Schema for updating a tool."""
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    description: Optional[str] = None
    vendor: Optional[str] = Field(None, max_length=100)
    website_url: Optional[HttpUrl] = None
    category_id: Optional[int] = Field(None, gt=0)
    monthly_cost: Optional[Decimal] = Field(None, ge=0, max_digits=10, decimal_places=2)
    owner_department: Optional[DepartmentType] = None
    status: Optional[ToolStatusType] = None


class ToolResponse(ToolBase):
    """Schema for tool response."""
    id: int
    status: ToolStatusType
    active_users_count: int
    category: Optional[str] = None  # Category name
    total_monthly_cost: Optional[Decimal] = None
    created_at: datetime
    updated_at: datetime
    usage_metrics: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class ToolListResponse(BaseModel):
    """Schema for list of tools response."""
    data: list[ToolResponse]
    total: int
    filtered: int
    filters_applied: Dict[str, Any]


# Tool Detail Response (for GET /api/tools/:id)
class ToolDetailResponse(ToolResponse):
    """Schema for detailed tool response."""
    pass
